src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java

```diff
@@ -0,0 +1,276 @@
+		setUndoPointAndRecordState( model, states );
+
+		// add tag set
+		TagSetStructure.TagSet tagset = TagSetUtils.addNewTagSetToModel( model, "tag set 1", Arrays.asList(
+				Pair.of( "tag1", Color.red.getRGB() ),
+				Pair.of( "tag2", Color.green.getRGB() ) ) );
+		TagSetStructure.Tag tag1 = tagset.getTags().get( 0 );
+		TagSetStructure.Tag tag2 = tagset.getTags().get( 1 );
+		TagSetUtils.tagSpot( model, tagset, tag1, a );
+		TagSetUtils.tagSpot( model, tagset, tag2, b );
+		TagSetUtils.tagLinks( model, tagset, tag1, Collections.singletonList( edge ) );
+		setUndoPointAndRecordState( model, states );
+
+		// change label
+		a.setLabel( "A0" );
+		b.setLabel( "B0" );
+		setUndoPointAndRecordState( model, states );
+
+		// change tag
+		TagSetUtils.tagSpot( model, tagset, tag2, a );
+		setUndoPointAndRecordState( model, states );
+
+		// change covariance
+		double[][] cov = { { 2, 0.1, 0 }, { 0.1, 2.5, 0 }, { 0, 0, 2.1 } };
+		a.setCovariance( cov );
+		setUndoPointAndRecordState( model, states );
+
+		// change spot position
+		b.setPosition( new double[] { 2.1, 2.3, 2.2 } );
+		setUndoPointAndRecordState( model, states );
+
+		// remove spot B
+		graph.remove( b );
+		setUndoPointAndRecordState( model, states );
+	}
+
+	public static void setUndoPointAndRecordState( Model model, List< String > states )
+	{
+		model.setUndoPoint();
+		states.add( modelAsString( model ) );
+	}
+
+	public static void undoAllRecordedPointsAndVerifyCorrectness( Model model, List< String > states )
+	{
+		for ( int i = states.size() - 1; i > 0; i-- )
+		{
+			assertEquals( states.get( i ), modelAsString( model ) );
+			model.undo();
+		}
+		assertEquals( states.get( 0 ), modelAsString( model ) );
+	}
+
+	public static void redoAllRecordedPointsAndVerifyCorrectness( Model model, List< String > states )
+	{
+		assertEquals( states.get( 0 ), modelAsString( model ) );
+		for ( int i = 1; i < states.size(); i++ )
+		{
+			model.redo();
+			assertEquals( states.get( i ), modelAsString( model ) );
+		}
+	}
+
+	private static String modelAsString( Model model )
+	{
+		return ModelUtils.dump( model, ModelUtils.DumpFlags.PRINT_TAGS );
+	}
+
+	/**
+	 * This test ensures that the undo and redo methods work as expected even
+	 * after the running ModelImporter.
+	 * <p>
+	 * The test first performs various changes to the model. Simply to fill the
+	 * undo-redo history with many different entries. Then running the ModelImporter
+	 * will reset the model and the undo-redo history.
+	 * This test ensures that the undo and redo methods work as expected even after
+	 * the model has been reset by the ModelImporter.
+	 */
+	@Test
+	public void testUndoRedoAfterModelImporter()
+	{
+		Model model = new Model();
+		// Fill the undo-redo history with many different entries:
+		makeVariousChangesAndRecordUndoPoints( model, new ArrayList<>() );
+		addFourSpotGraph( model );
+		// Reset the model by running the ModelImporter / Simulate an import operation:
+		new ModelImporter( model )
```

Comment:
This is difficult to read and understand what is happening and why it is written in this way.

However, the constructor of ModelImporter is protected and should not be made public only for this test thus I do not see a simpler alternative to this.

Reply:
There is now a Importer subclass that avoids this problem.

Reply:
Good idea. I like it.